## As perguntas do CEO da House Rocket

1 - Quantas casas estão disponíveis para compra?

2 - Quantos atributos (nº de quartos, vagas de garagem, m², varanda, etc) as casas possuem?

3 - Quais são os atributos?

4 - Qual a casa mais cara do portfólio? (caso com maior valor)

5 - Qual a casa com maior número de quarto?



## Planejmaneto do produto final

(O que vamos entregar para o CEO?)

**1 - Planejamento do PRODUTO FINAL**

* O que vou entregar? (Planilha, texto, email, modelo de ML...)

R: texto com as respostas

* Como vai ser a entregar? (Quais colunas, quais linhas, vai ter gráfico?)

R: perguntas | respostas


**2 - Planejamento do PROCESSO**

* Onde está as informações? (Excel, BD, API...)

R: kc_house_data

* Como coletar as informações? (SQL, Python, Stramlit...)

R: utilizar o dataset disponibilizado pelo CEO

* Responder as perguntas

R: utilizar o dataset disponibilizado pelo CEO para responder

**3 - Planejamento das FERRAMENTAS**

* Quais ferramentas usar?

R: Excel e Python

In [1]:
# Carregando os dados do estudo de caso em um jupyter notebook

## importar pandas

import pandas as pd

In [2]:
# importando o dataset usando o método pd.read_excel

df = pd.read_csv('kc_house_data.csv')

In [3]:
# cinco primeiras linhas do dataset

df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [4]:
# cinco últimas linhas

df.tail()

id             date     price  bedrooms  bathrooms  \
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[5 rows x 21 columns]

Cada linha representa um imóvel do dataset. Nas saídas acima, vemos as 5 primerias e as 5 últimas colunas. Todos os imóveis estão identificados por um **id**.

A coluna **id** parece conter identificadores exclusivos.

Para verificar se isso ocorre realmente em todo o dataset, podemos contar o número de valores exclusivos usando o método **.nunique()** da série (ou seja, coluna) **id**.

Primeiro, selecionaremos a coluna usando colchetes.

In [5]:
# descobrindo se há problema de qualidade de dados
# A função pandas .nunique() retorna o número de elementos únicos no objeto.

df['id'].nunique()

21436

Agora, vamos executar o comando **.shape** para descobrir o número de linhas do dataset

In [6]:
# descobrindo o números de linhas do dataset

df.shape

(21613, 21)

Identificamos que o número de **id**s exclusivos é menor do que o número de linhas. Isso significa que o **id** não é um identificador exclusivo para todas as linhas do dataset.

Pode estar ocorrendo duplicação de **id**s. Se está hipótese estiver correta, em que quantidades está ocorrendo estas duplicações? Um único **id** está sendo duplicado várias vezes? Quantos **id**s estão sendo duplicados?

Podemos usar o método **.value_counts()** na série **id** para começar a responder a essas perguntas.

Ele é semelhante a um procedimento group by/count em SQL e listará os **id**s exclusivos e a frequência com que ocorrem.

Executaremos essa operação na próxima etapa e armazenaremos as contagens de valores em uma variável id_counts.

In [11]:
# armazene as contagens de valores em uma variável definida como id_counts e exiba os valores armazenados usando o método .head()

id_counts = df['id'].value_counts()

In [12]:
id_counts.head()

795000620     3
2206700215    2
643300040     2
3333002450    2
1995200200    2
Name: id, dtype: int64

Observe que o **.head()** retorna, por padrão, as cinco primeiras linhas. Podemos especificar o número de itens a serem exibidos passando o número desejado nos parênteses, ().

In [9]:
id_counts.head(20)

795000620     3
2206700215    2
643300040     2
3333002450    2
1995200200    2
1781500435    2
3904100089    2
3323059027    2
6300000226    2
9809000020    2
723049156     2
1254200015    2
2560801222    2
6669020290    2
8651402750    2
3969300030    2
1237500540    2
4154300296    2
4302201085    2
1450100390    2
Name: id, dtype: int64

In [10]:
# exiba o número de entradas duplicadas agrupadas executando outra contagem de valores

id_counts.value_counts()

1    21260
2      175
3        1
Name: id, dtype: int64

Na saída acima podemos ver que a maioria dos **id**s (21260) ocorre exatamente uma única vez, como esperado.

No entanto, 175 **id**s ocorrem duas vezes e 1 **id** ocorre 3 vezes. Logo, nenhum **id** ocorre mais do que três vezes.

De posse dessas informações, estamos prontos para começar a examinar mais detalhadamente esse problema de qualidade de dados e corrigi-lo?

Podemos criaremos máscaras booleanas para limpar melhor os dados?

## O que fazer? Corrigir esses equivocos já ou esperar as próximas aulas??

# Repondendo o CEO

### 1 - Quantas casas estão disponíveis para compra?

Como descrito acima, cada linha representa um imóvel. Então, uma forma prática e rápida de verificar a quatidade de linhas e de colunas é utilizando a função **.shape**.

In [11]:
# tamanho de linhas e colunas

df.shape

(21613, 21)

Acima recebemos que o dataset contém 21613 linhas e 21 colunas. 

A resposta para a primeira perguta está aqui, **existem 21613 imóveis à venda**

In [13]:
# Estatísticas descritivas básicas para cada coluna (ou GroupBy)

df.describe() 

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

### 2 - Quantos atributos (nº de quartos, vagas de garagem, m², varanda, etc) as casas possuem?

Uma maneira concisa de acessar as colunas é chamando a função **.columns**.

In [12]:
df.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [14]:
len(df.columns)

21

Acima, contamos quantas colunas existem no dataset. Quais são os elementos que podem ser apresentados como atributos?

- 'id' - código de identificação dos imóveis, 
- 'date' - data (do que?), 
- 'price' - preço/valor, 
- 'bedrooms' - quartos, 
- 'bathrooms' - banheiros, 
- 'sqft_living' - m² (construído?),
- 'sqft_lot' - m² (do terreno?), 
- 'floors' - pavimentos, 
- 'waterfront', 
- 'view' - vista (pra onde?), 
- 'condition' - estado, 
- 'grade' - ,
- 'sqft_above' - , 
- 'sqft_basement' - , 
- 'yr_built' - ano de construção, 
- 'yr_renovated' - ano de reforma, 
- 'zipcode' - CEP,
- 'lat' - latidude, 
- 'long' - longitude, 
- 'sqft_living15' - m² (repetido), 
- 'sqft_lot15' - m² (repetido).



### 4 - Qual a casa mais cara do portfólio? (caso com maior valor)

Para responder esta pergunta podemos usar a função **df.sort_values('mpg')**. **.sort_values()** ordena as linhas pelos valores de uma coluna (do menor ao maior)

In [17]:
# Para facilitar a visualização, iremos selecior apenas as colunas 'id' e 'price' abrindo 
# duplo conchete e declarando as colunas desejadas

df[['id', 'price']].sort_values('price', ascending=False)

id      price
7252   6762700020  7700000.0
3914   9808700762  7062500.0
9254   9208900037  6885000.0
4411   2470100110  5570000.0
1448   8907500070  5350000.0
...           ...        ...
8274   3883800011    82000.0
16198  3028200080    81000.0
465    8658300340    80000.0
15293    40000362    78000.0
1149   3421079032    75000.0

[21613 rows x 2 columns]

Repondendo a pergunta 4. **O imóvel mais valiodo do portfólio da empresa é o imóvel 7252 que custa $7,700,000.0**

### 5 - Qual a casa com maior número de quarto?

Podemos utilizar a mesma lógica aplicada acima para responder está questão.

In [20]:
df[['id', 'bedrooms']].sort_values('bedrooms', ascending=False)

id  bedrooms
15870  2402100895        33
8757   1773100755        11
15161  5566100170        10
13314   627300145        10
19254  8812401450        10
...           ...       ...
19452  3980300371         0
8484   2310060040         0
875    6306400140         0
8477   2569500210         0
9773   3374500520         0

[21613 rows x 2 columns]

Repondendo a pergunta 4. **O imóvel com mais números de quartos do portfólio da empresa é o imóvel 15870 com 33 quartos**

# Resumo

Escreva o resumo aqui (perguntas | respostas)

## TAREFAS

Novas perguntas do CEO

- Conte a quantidade de linhas do dataset;
- Conte a quantidade de colunas do dataset;
- Mostre o nome das colunas do dataset;
- Ordene o dataset pelo preço das casas;